pkg/commands/artifact/k8s.go

```diff
@@ -0,0 +1,274 @@
+	k8sArtifacts, err := trivyk8s.ListArtifacts(ctx.Context)
+	if err != nil {
+		return xerrors.Errorf("get k8s artifacts error: %w", err)
+	}
+
+	report, err := k8sRun(ctx, opt, cacheClient, k8sArtifacts)
+	if err != nil {
+		return xerrors.Errorf("k8s scan error: %w", err)
+	}
+	report.ClusterName = cluster.GetCurrentContext()
+
+	if err = k8sReport.Write(report, pkgReport.Option{
+		Format: "json", // for now json is the default
+		Output: opt.Output,
+	}); err != nil {
+		return xerrors.Errorf("unable to write results: %w", err)
+	}
+
+	exit(opt, report.Failed())
+
+	return nil
+}
+
+func k8sRun(cliContext *cli.Context, opt Option, cacheClient cache.Cache, k8sArtifacts []*artifacts.Artifact) (k8sReport.Report, error) {
+	ctx, cancel := context.WithTimeout(cliContext.Context, opt.Timeout)
+	defer cancel()
+
+	// progress bar
+	bar := pb.StartNew(len(k8sArtifacts))
+	if opt.NoProgress {
+		bar.SetWriter(io.Discard)
+	}
+	defer bar.Finish()
+
+	// image scanner configurations
+	imageScannerConfig, imageScannerOptions, err := initImageScannerConfig(ctx, opt, cacheClient)
+	if err != nil {
+		return k8sReport.Report{}, xerrors.Errorf("scanner config error: %w", err)
+	}
+
+	// config scanner configurations
+	configScannerConfig, configScannerOptions, err := initConfigScannerConfig(ctx, opt, cacheClient)
+	if err != nil {
+		return k8sReport.Report{}, xerrors.Errorf("scanner config error: %w", err)
+	}
+
+	vulns := make([]k8sReport.Resource, 0)
+	misconfigs := make([]k8sReport.Resource, 0)
+
+	// Loops once over all artifacts, and execute scanners as necessary. Not every artifacts has an image,
+	// so image scanner is not always executed.
+	for _, artifact := range k8sArtifacts {
+		bar.Increment()
+
+		// scan images if present
+		for _, image := range artifact.Images {
+			imageReport, err := k8sScan(ctx, image, imageScanner, imageScannerConfig, imageScannerOptions)
+			if err != nil {
+				// add error to report
+				log.Logger.Debugf("failed to scan image %s: %s", image, err)
+				vulns = append(vulns, newK8sResource(artifact, imageReport, err))
+				continue
+			}
+
+			imageReport, err = filter(ctx, opt, imageReport)
+			if err != nil {
+				return k8sReport.Report{}, xerrors.Errorf("filter error: %w", err)
+			}
+
+			vulns = append(vulns, newK8sResource(artifact, imageReport, nil))
+		}
+
+		// scan configurations
+		configReport, err := k8sScanConfig(ctx, configScannerConfig, configScannerOptions, artifact)
+		if err != nil {
+			// add error to report
+			log.Logger.Debugf("failed to scan config %s/%s: %s", artifact.Kind, artifact.Name, err)
+			misconfigs = append(misconfigs, newK8sResource(artifact, configReport, err))
+		}
+
+		configReport, err = filter(ctx, opt, configReport)
+		if err != nil {
+			return k8sReport.Report{}, xerrors.Errorf("filter error: %w", err)
+		}
+
+		misconfigs = append(misconfigs, newK8sResource(artifact, configReport, nil))
+	}
+
+	return k8sReport.Report{
+		SchemaVersion:     0,
```

Comment:
The version 0 looks weird. Is this because this feature is still experimental?

Reply:
That was the idea, but I can change it to 1 if you think it makes more sense.

Reply:
It is ok then. I agree the schema is not fixed yet. I'm sure we will fix the schema soon.